server/src/main/java/org/elasticsearch/action/support/CountDownActionListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.action.ActionListener;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Wraps another listener and adds a counter -- each invocation of this listener will decrement the counter, and when the counter has been
+ * exhausted the final invocation of this listener will delegate to the wrapped listener. Similar to {@link GroupedActionListener}, but for
+ * the cases where tracking individual results is not useful.
+ */
+public final class CountDownActionListener extends ActionListener.Delegating<Void, Void> {
+
+    private final AtomicInteger countDown;
+    private final AtomicReference<Exception> failure = new AtomicReference<>();
+
+    /**
+     * Creates a new listener
+     * @param groupSize the group size
+     * @param delegate the delegate listener
+     */
+    public CountDownActionListener(int groupSize, ActionListener<Void> delegate) {
+        super(Objects.requireNonNull(delegate));
+        if (groupSize <= 0) {
+            assert false : "illegal group size [" + groupSize + "]";
+            throw new IllegalArgumentException("groupSize must be greater than 0 but was " + groupSize);
+        }
+        countDown = new AtomicInteger(groupSize);
+    }
+
+    /**
+     * Creates a new listener
+     * @param groupSize the group size
+     * @param runnable the runnable
+     */
+    public CountDownActionListener(int groupSize, Runnable runnable) {
+        this(groupSize, ActionListener.wrap(Objects.requireNonNull(runnable)));
+    }
+
+    private boolean countDown() {
+        final var result = countDown.getAndUpdate(current -> Math.max(0, current - 1));
+        assert result > 0;
+        return result == 1;
+    }
+
+    @Override
+    public void onResponse(Void element) {
+        if (countDown()) {
+            if (failure.get() != null) {
+                super.onFailure(failure.get());
+            } else {
+                delegate.onResponse(element);
+            }
+        }
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        if (failure.compareAndSet(null, e) == false) {
+            failure.accumulateAndGet(e, (current, update) -> {
+                // we have to avoid self-suppression!
+                if (update != current) {
+                    current.addSuppressed(update);
+                }
+                return current;
+            });
+        }
+        if (countDown()) {
```

Comment:
I like keeping it as waiting for all dependents -- in the simple case of only a single failure, one could argue that it's wasting time, but it does collect up multiple failures and report them all via the `addSuppressed` call, and we can't be sure that all the rest of the calls would succeed. That said, though, mostly I'm just going for parity with what `GroupedActionListener` already does here, and it doesn't fail-fast, so I figured I shouldn't either.

Reply:
:+1: from me for not failing fast here. It's easy enough for callers to make it fail-fast if needed:

        final var runOnceListener = ActionListener.notifyOnce(listener);
        final var cdal = new CountDownActionListener(count, runOnceListener).delegateResponse((l, e) -> runOnceListener.onFailure(e));

I opened https://github.com/elastic/elasticsearch/issues/92451 because I think we can make this a little more heap-efficient too.

OTOH if we made the implementation fail-fast then a caller wanting to wait for completion could do a similar trick but would have to use a separate countdown, and that kind of defeats the point of this class 😄

Reply:
Sounds good